{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform feature selection and classification to predict the gene status of the gliomas patients.\n",
    "\n",
    "For example, we can use this code and the radiomic features to predict:\n",
    "- 1. LGG vs. GBM;\n",
    "- 2. IDH mutant vs. IDH wildtype;\n",
    "- 3. 1p/19q codeleted vs. 1p/19q intact;\n",
    "- 4. MGMT methylated vs. MGMT unmethylated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    " \n",
    "from sklearn.model_selection import train_test_split\n",
    "from mySettings import get_basic_settings\n",
    "from classificationUtils import perform_binary_classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_classification_tasks_dict(experiment_class, task_name):\n",
    "    \n",
    "    if experiment_class==\"BraTS2021\":\n",
    "        classification_tasks_dict=get_classification_tasks_dict_BraTS2021(task_name)\n",
    "\n",
    "    elif experiment_class==\"TCGA_IDH\" or experiment_class==\"TCGA_MGMT\":\n",
    "        classification_tasks_dict=get_classification_tasks_dict_TCGA(experiment_class, task_name) \n",
    "        \n",
    "    else:\n",
    "        raise Exception(\"Undefined experiment class: {}.\".format(experiment_class))\n",
    "        \n",
    "    return classification_tasks_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define classification settings for BraTS2021 dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def convert_complex_to_real(datadf, feature_namelist):\n",
    "    # convert_complex_to_real\n",
    "    datadf_converted=datadf.copy()\n",
    "    datadf_converted[feature_namelist]=datadf[feature_namelist].applymap(lambda x: complex(x).real)\n",
    "    \n",
    "    return datadf_converted\n",
    "\n",
    "\n",
    "def get_feature_columns(train_data, feature_filter_dict):\n",
    "    \"\"\"\n",
    "    Filter features according to modalities, image types, and tumor subregions.\n",
    "    Note that, for the same subregion, the shape features should be the same for different modalities and image types.\n",
    "    \n",
    "    Feature_filter_dict has three keys:\n",
    "    - modality_list: [\"t1\", \"t1ce\"/\"t1Gd\", \"t2\", \"flair\"];\n",
    "    - imageType_list: [\"original\", \"gradient\", \"log-sigma-1-0-mm-3D\", \"log-sigma-3-0-mm-3D\"];\n",
    "    - tumor_subregion_list: [\"NCR\", \"ED\", \"ET\", \"TC\", \"wholeTumor\"]\n",
    "    \"\"\"\n",
    "    \n",
    "    modality_list=feature_filter_dict[\"modality_list\"]\n",
    "    imageType_list=feature_filter_dict[\"imageType_list\"]\n",
    "    tumor_subregion_list=feature_filter_dict[\"tumor_subregion_list\"]\n",
    "    \n",
    "    all_columns=train_data.columns\n",
    "           \n",
    "    #1. filter the shape features. \n",
    "    # - Different modalities have the same shape features. \n",
    "    # - The shape features are always saved in \"original\" image type.\n",
    "    shape_feature_columns=[]\n",
    "    for column in all_columns:\n",
    "        column_name_split=column.split(\"_\")\n",
    "        example_modality_name=modality_list[0]\n",
    "        if column_name_split[0]==example_modality_name and (column_name_split[1] in tumor_subregion_list) and column_name_split[2]==\"original\" and column_name_split[3]==\"shape\":\n",
    "            shape_feature_columns.append(column)\n",
    "       \n",
    "    # 2. filter other features (not include shape features) from different modalities, different image types, and different subregions;\n",
    "    other_feature_columns=[]\n",
    "    for column in all_columns:\n",
    "        column_name_split=column.split(\"_\")\n",
    "        if (column_name_split[0] in modality_list) and (column_name_split[1] in tumor_subregion_list) and (column_name_split[2] in imageType_list) and (column_name_split[3]!=\"shape\"): \n",
    "            other_feature_columns.append(column)\n",
    "                \n",
    "    # connect the shape features and other features.\n",
    "    final_feature_columns=shape_feature_columns+other_feature_columns\n",
    "    print(\"There are {} radiomic features in total, including {} shape features and {} other features!\".format(len(final_feature_columns), len(shape_feature_columns), len(other_feature_columns)))\n",
    "    \n",
    "    return final_feature_columns\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#======================================= BraTS2021 classification task settings =======================================================\n",
    "\"\"\"\n",
    "Perform MGMT classification for BraTS2021 competition.\n",
    "\"\"\"\n",
    "def get_classification_tasks_dict_BraTS2021(task_name):\n",
    "    basepath=\"G://PhDProjects/RadiogenomicsProjects/BraTS2021\"\n",
    "    basic_settings=get_basic_settings() \n",
    "    feature_filter_dict=basic_settings[\"feature_filter_dict\"]\n",
    "    imageType_list=feature_filter_dict[\"imageType_list\"][0]\n",
    "    harmonization_detail= \"withoutComBat\" if basic_settings[\"harmonization_method\"]==\"withoutComBat\" else basic_settings[\"harmonization_method\"]+\"_\"+basic_settings[\"harmonization_label\"]+\"_refbat-\"+str(basic_settings[\"harmonization_ref_batch\"])\n",
    "    base_results_path=basepath+\"/Results/\"+imageType_list+\"/Results_BraTS2021_MGMT/\"+harmonization_detail+\"-\"+basic_settings[\"imbalanced_data_strategy\"]\n",
    "    classification_tasks_dict={}\n",
    "    \n",
    "    \n",
    "    ## basic excel path settings\n",
    "    classification_tasks_dict[\"BraTS2021_5.101_segNiiData_base\"]={\n",
    "        \"train_excel_path\": basepath+\"/Features/final_metadata/features_BraTS2021_train.xlsx\",\n",
    "        \"test_excel_path_dict\": {\"validation_data\": basepath+\"/Features/final_metadata/features_BraTS2021_validation.xlsx\"},\n",
    "    }\n",
    "    \n",
    "#     classification_tasks_dict[\"BraTS2021_5.201_dcmToNiiData_base\"]={\n",
    "#         \"train_excel_path\": basepath+\"/Features/final_metadata/features_BraTS2021_train_dcm_to_nii.xlsx\",\n",
    "#         \"test_excel_path_dict\": {\"validation_data\": basepath+\"/Features/final_metadata/features_BraTS2021_validation_dcm_to_nii.xlsx\"},\n",
    "#     }\n",
    "    \n",
    "    #Add other settings (\"train_data\", \"test_data_dict\", \"label_column\", \"base_results_path\", \"feature_columns\") for each task. \n",
    "    if task_name is not None:\n",
    "        find_task=False\n",
    "        for task, classification_settings in classification_tasks_dict.items():\n",
    "            if task_name==task:\n",
    "                find_task=True\n",
    "                train_excel_path=classification_settings[\"train_excel_path\"]\n",
    "                test_excel_path_dict=classification_settings[\"test_excel_path_dict\"]\n",
    "\n",
    "                # preprocessing train data\n",
    "                train_data=pd.read_excel(train_excel_path, index_col=0)\n",
    "                feature_columns=get_feature_columns(train_data, feature_filter_dict)\n",
    "                train_data=preprocessing_data(train_data, feature_columns)\n",
    "\n",
    "                # preprocessing test data\n",
    "                test_data_dict={}\n",
    "                for description, test_excel_path in test_excel_path_dict.items():\n",
    "                    test_data=pd.read_excel(test_excel_path, index_col=0)\n",
    "                    test_data=preprocessing_data(test_data, feature_columns)\n",
    "                    test_data_dict[description]=test_data\n",
    "\n",
    "                # set and save the settings\n",
    "                classification_settings[\"train_data\"]=train_data\n",
    "                classification_settings[\"test_data_dict\"]=test_data_dict\n",
    "                classification_settings[\"label_column\"]=\"MGMT_value\" \n",
    "                classification_settings[\"base_results_path\"]=base_results_path\n",
    "                classification_settings[\"feature_columns\"]=feature_columns\n",
    "                classification_tasks_dict[task]=classification_settings\n",
    "\n",
    "        if not find_task:\n",
    "            raise Exception(\"Task {} not defined!\".format(task_name))\n",
    "        \n",
    "    return classification_tasks_dict                       \n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define settings for TCGA dataset, for predicting tumor grade, IDH and 1p/19q, and for predicting MGMT."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\"\"\"\n",
    "For the public features of TCGA dataset, classify the columns into features, interested clinical info and the classification targets.\n",
    "\"\"\"\n",
    "def get_column_list_for_TCGA(columns):\n",
    "\n",
    "    feature_namelist=[]\n",
    "    clinical_namelist=[]\n",
    "    classification_label_namelist=[]\n",
    "    ET_related_feature_namelist=[]\n",
    "    for column in columns:\n",
    "        if column.startswith(\"ET_related_feature_\"): \n",
    "            ET_related_feature_namelist.append(column)\n",
    "            \n",
    "        elif column in [\"age\", \"is_female\"]:\n",
    "            clinical_namelist.append(column)\n",
    "\n",
    "        elif column in [\"is_GBM\", \"is_IDH_mutant\", \"is_1p19q_codeleted\", \"is_MGMT_Methylated\"]:\n",
    "            classification_label_namelist.append(column)\n",
    "\n",
    "        else:\n",
    "            column_prefix=column.split(\"_\")[0]\n",
    "            if column_prefix in [\"VOLUME\", \"DIST\", \"INTENSITY\", \"HISTO\",\"SPATIAL\", \"ECCENTRICITY\", \"SOLIDITY\", \"TEXTURE\", \"TGM\"]:\n",
    "                feature_namelist.append(column)\n",
    "                \n",
    "    feature_column_list=feature_namelist+ET_related_feature_namelist\n",
    "    print(\"\\n len(feature_column_list)={}\".format(len(feature_column_list)))\n",
    "    print(\"\\n clinical_namelist={}\".format(clinical_namelist))\n",
    "    print(\"\\n classification_label_namelist={}\".format(classification_label_namelist))\n",
    "    \n",
    "    return feature_column_list, clinical_namelist, classification_label_namelist \n",
    "\n",
    "\n",
    "#========== classification task settings for TCGA datast: GBM, IDH and 1p/19q, and MGMT =========================\n",
    "def get_classification_tasks_dict_TCGA(experiment_class, task_name):\n",
    "    # basic paths\n",
    "    basepath=\"G://PhDProjects/RadiogenomicsProjects/GliomasSubtypes\" \n",
    "    basic_settings=get_basic_settings() \n",
    "    features_for_TCGA=basic_settings[\"features_for_TCGA\"]\n",
    "    normalization_method=\"\" if features_for_TCGA==\"public_features\" else basic_settings[\"normalization_method\"]\n",
    "    feature_filter_dict=basic_settings[\"feature_filter_dict\"]\n",
    "    random_seed=basic_settings[\"random_seed\"]\n",
    "    imageType_list=feature_filter_dict[\"imageType_list\"][0]\n",
    "    harmonization_detail= \"withoutComBat\" if basic_settings[\"harmonization_method\"]==\"withoutComBat\" else basic_settings[\"harmonization_method\"]+\"_\"+basic_settings[\"harmonization_label\"]+\"_refbat-\"+str(basic_settings[\"harmonization_ref_batch\"])\n",
    "    base_results_path=basepath+\"/Results_randomseed\"+str(random_seed)+\"/\"+imageType_list+\"/\"+experiment_class+\"_\"+features_for_TCGA+\"_\"+normalization_method+\"/\"+harmonization_detail+\"-\"+basic_settings[\"imbalanced_data_strategy\"]\n",
    "    \n",
    "    \n",
    "    ##--------------------- Prepare the data -------------------\n",
    "    # data path\n",
    "    if experiment_class==\"TCGA_IDH\" and features_for_TCGA==\"extracted_features\":\n",
    "        train_data_excel_path=os.path.join(basepath, \"Features\", \"final_metadata\", normalization_method, \"TCGA_extracted_features_IDH_train.xlsx\") \n",
    "        test_data_excel_path=os.path.join(basepath, \"Features\", \"final_metadata\", normalization_method, \"TCGA_extracted_features_IDH_test.xlsx\")\n",
    "        \n",
    "    elif experiment_class==\"TCGA_IDH\" and features_for_TCGA==\"public_features\":\n",
    "        train_data_excel_path=os.path.join(basepath, \"Features\", \"final_metadata\", \"TCGA_public_features_IDH_train.xlsx\") \n",
    "        test_data_excel_path=os.path.join(basepath, \"Features\", \"final_metadata\", \"TCGA_public_features_IDH_test.xlsx\")\n",
    "    \n",
    "    elif experiment_class==\"TCGA_MGMT\" and features_for_TCGA==\"extracted_features\":\n",
    "        train_data_excel_path=os.path.join(basepath, \"Features\", \"final_metadata\", normalization_method, \"TCGA_extracted_features_MGMT_train.xlsx\") \n",
    "        test_data_excel_path=os.path.join(basepath, \"Features\", \"final_metadata\", normalization_method,  \"TCGA_extracted_features_MGMT_test.xlsx\")\n",
    "        \n",
    "    elif experiment_class==\"TCGA_MGMT\" and features_for_TCGA==\"public_features\":\n",
    "        train_data_excel_path=os.path.join(basepath, \"Features\", \"final_metadata\", \"TCGA_public_features_MGMT_train.xlsx\") \n",
    "        test_data_excel_path=os.path.join(basepath, \"Features\", \"final_metadata\", \"TCGA_public_features_MGMT_test.xlsx\")\n",
    "        \n",
    "    else:\n",
    "        raise Exception(\"Unfined feature type for TCGA dataset!\")\n",
    "\n",
    "        \n",
    "    ## ==================== Resplit the data with stratified 1p/19q codeletion status. ============================\n",
    "    if experiment_class==\"TCGA_IDH\":\n",
    "        stratify_column=\"is_1p19q_codeleted\" \n",
    "    elif experiment_class==\"TCGA_MGMT\":\n",
    "        stratify_column=\"is_MGMT_Methylated\" \n",
    "        \n",
    "    train_data=pd.read_excel(train_data_excel_path, index_col=0)\n",
    "    test_data=pd.read_excel(test_data_excel_path, index_col=0)\n",
    "    data=pd.concat([train_data, test_data], axis=0) \n",
    "    train_data, test_data, y_train, y_test = train_test_split(data, data[stratify_column], stratify= data[stratify_column], \n",
    "                                                              test_size=0.3, random_state=random_seed)\n",
    "    \n",
    "    train_data_excel_path=train_data_excel_path[:-5]+\"_resplited_randomseed_\"+str(random_seed)+\".xlsx\"\n",
    "    test_data_excel_path=test_data_excel_path[:-5]+\"_resplited_randomseed_\"+str(random_seed)+\".xlsx\"\n",
    "    train_data.to_excel(train_data_excel_path)\n",
    "    test_data.to_excel(test_data_excel_path)\n",
    "    print(\"\\n\\n======Resplit the train and test data, stratified by {}, random_seed={}=======\".format(stratify_column, random_seed))\n",
    "    print(\"-Before split: data.shape={}; \\After split, train_data.shape={}, test_data.shape={}\".format(data.shape, train_data.shape, test_data.shape))\n",
    "    #====================================================================================\n",
    "    \n",
    "    # test data        \n",
    "    test_excel_path_dict={\"test_data\": test_data_excel_path}\n",
    "    \n",
    "    ## === preprocess the train data===\n",
    "    #read train data\n",
    "    train_data=pd.read_excel(train_data_excel_path, index_col=0)\n",
    "    \n",
    "    #get different kinds of feature columns;\n",
    "    if features_for_TCGA==\"public_features\":\n",
    "        feature_namelist, clinical_namelist, classification_label_namelist=get_column_list_for_TCGA(train_data)\n",
    "    else:\n",
    "        _, clinical_namelist, classification_label_namelist=get_column_list_for_TCGA(train_data)\n",
    "        feature_namelist=get_feature_columns(train_data, feature_filter_dict)\n",
    "    \n",
    "    #preprocess train data.\n",
    "    train_data=preprocessing_data(train_data, feature_namelist)\n",
    "    train_data_LGG=train_data[train_data[\"is_GBM\"] == 0] # Filter the LGG data for predicting 1p/19q status.\n",
    "    train_data_GBM=train_data[train_data[\"is_GBM\"] == 1] # Filter the GBM data for predicting MGMT methylation status.\n",
    "    \n",
    "    ## ==== preprocess the test dataset ===\n",
    "    test_data_dict={}\n",
    "    test_data_LGG_dict={}\n",
    "    test_data_GBM_dict={}\n",
    "    for description, test_excel_path in test_excel_path_dict.items():\n",
    "        test_data=pd.read_excel(test_excel_path, index_col=0)\n",
    "        test_data=preprocessing_data(test_data, feature_namelist)\n",
    "        test_data_LGG=test_data[test_data[\"is_GBM\"] == 0]\n",
    "        test_data_GBM=test_data[test_data[\"is_GBM\"] == 1]\n",
    "        test_data_dict[description]=test_data\n",
    "        test_data_LGG_dict[description]=test_data_LGG\n",
    "        test_data_GBM_dict[description]=test_data_GBM\n",
    "    \n",
    "   \n",
    "    #-------------------------- Define the classification tasks ----------------------------------\n",
    "    classification_tasks_dict={}\n",
    "    if experiment_class==\"TCGA_IDH\":\n",
    "        #========= predict LGG vs. GBM  ========\n",
    "        classification_tasks_dict[\"TCGA_1.101_isGBM_base\"]={\n",
    "            \"train_data\": train_data, \n",
    "            \"test_data_dict\": test_data_dict, \n",
    "            \"feature_columns\":feature_namelist, \n",
    "            \"keep_feature_columns\": [],\n",
    "            \"label_column\":\"is_GBM\",\n",
    "            \"base_results_path\":base_results_path}\n",
    "\n",
    "        classification_tasks_dict[\"TCGA_1.102_isGBM_withClinicalInfo\"]={\n",
    "            \"train_data\": train_data, \n",
    "            \"test_data_dict\": test_data_dict, \n",
    "            \"feature_columns\":feature_namelist,\n",
    "            \"keep_feature_columns\": clinical_namelist,\n",
    "            \"label_column\":\"is_GBM\",\n",
    "            \"base_results_path\":base_results_path}\n",
    "\n",
    "\n",
    "        #========= predict IDH mutation status  ========= \n",
    "        classification_tasks_dict[\"TCGA_2.101_isIDHMutant_base\"]= {\n",
    "            \"train_data\": train_data, \n",
    "            \"test_data_dict\": test_data_dict, \n",
    "            \"feature_columns\":feature_namelist, \n",
    "            \"keep_feature_columns\": [],\n",
    "            \"label_column\":\"is_IDH_mutant\",\n",
    "            \"base_results_path\":base_results_path}\n",
    "\n",
    "        classification_tasks_dict[\"TCGA_2.102_isIDHMutant_withClinicalInfo\"]={\n",
    "            \"train_data\": train_data, \n",
    "            \"test_data_dict\": test_data_dict, \n",
    "            \"feature_columns\":feature_namelist,\n",
    "            \"keep_feature_columns\": clinical_namelist,\n",
    "            \"label_column\":\"is_IDH_mutant\",\n",
    "            \"base_results_path\":base_results_path}\n",
    "\n",
    "        #========= predict 1p/19q codeletion status ========= \n",
    "        classification_tasks_dict[\"TCGA_3.101_is1p19qCodeleted_base\"]={\n",
    "            \"train_data\": train_data, \n",
    "            \"test_data_dict\": test_data_dict, \n",
    "            \"feature_columns\":feature_namelist, \n",
    "            \"keep_feature_columns\": [],\n",
    "            \"label_column\":\"is_1p19q_codeleted\",\n",
    "            \"base_results_path\":base_results_path}\n",
    "\n",
    "        classification_tasks_dict[\"TCGA_3.102_is1p19qCodeleted_withClinicalInfo\"]={\n",
    "            \"train_data\": train_data, \n",
    "            \"test_data_dict\": test_data_dict, \n",
    "            \"feature_columns\":feature_namelist,\n",
    "            \"keep_feature_columns\": clinical_namelist,\n",
    "            \"label_column\":\"is_1p19q_codeleted\",\n",
    "            \"base_results_path\":base_results_path}\n",
    "        \n",
    "        classification_tasks_dict[\"TCGA-LGG_3.201_is1p19qCodeleted_base\"]={\n",
    "            \"train_data\": train_data_LGG, \n",
    "            \"test_data_dict\": test_data_LGG_dict, \n",
    "            \"feature_columns\":feature_namelist, \n",
    "            \"keep_feature_columns\": [],\n",
    "            \"label_column\":\"is_1p19q_codeleted\",\n",
    "            \"base_results_path\":base_results_path}\n",
    "\n",
    "        classification_tasks_dict[\"TCGA-LGG_3.202_is1p19qCodeleted_withClinicalInfo\"]={\n",
    "            \"train_data\": train_data_LGG, \n",
    "            \"test_data_dict\": test_data_LGG_dict, \n",
    "            \"feature_columns\":feature_namelist, \n",
    "            \"keep_feature_columns\": clinical_namelist,\n",
    "            \"label_column\":\"is_1p19q_codeleted\",\n",
    "            \"base_results_path\":base_results_path}\n",
    "        \n",
    "    elif experiment_class==\"TCGA_MGMT\":\n",
    "        #========= predict MGMT methylated vs. unmethylated for LGG and GBM data ========\n",
    "        classification_tasks_dict[\"TCGA_4.101_isMGMTMethylated_base\"]={\n",
    "            \"train_data\": train_data, \n",
    "            \"test_data_dict\": test_data_dict, \n",
    "            \"feature_columns\":feature_namelist,\n",
    "            \"keep_feature_columns\": [],\n",
    "            \"label_column\":\"is_MGMT_Methylated\",\n",
    "            \"base_results_path\":base_results_path}\n",
    "\n",
    "        classification_tasks_dict[\"TCGA_4.102_isMGMTMethylated_withClinicalInfo\"]={\n",
    "            \"train_data\": train_data, \n",
    "            \"test_data_dict\": test_data_dict, \n",
    "            \"feature_columns\":feature_namelist,\n",
    "            \"keep_feature_columns\": clinical_namelist,\n",
    "            \"label_column\":\"is_MGMT_Methylated\",\n",
    "            \"base_results_path\":base_results_path}\n",
    "\n",
    "        #========= predict MGMT methylated vs. unmethylated for GBM data ========\n",
    "        classification_tasks_dict[\"TCGA-GBM_4.201_isMGMTMethylated_base\"]={\n",
    "            \"train_data\": train_data_GBM, \n",
    "            \"test_data_dict\": test_data_GBM_dict, \n",
    "            \"feature_columns\":feature_namelist, \n",
    "            \"keep_feature_columns\": [],\n",
    "            \"label_column\":\"is_MGMT_Methylated\",\n",
    "            \"base_results_path\":base_results_path}\n",
    "\n",
    "        classification_tasks_dict[\"TCGA-GBM_4.202_isMGMTMethylated_withClinicalInfo\"]={\n",
    "            \"train_data\": train_data_GBM, \n",
    "            \"test_data_dict\": test_data_GBM_dict, \n",
    "            \"feature_columns\":feature_namelist,\n",
    "            \"keep_feature_columns\": clinical_namelist,\n",
    "            \"label_column\":\"is_MGMT_Methylated\",\n",
    "            \"base_results_path\":base_results_path}\n",
    "        \n",
    "         #========= predict MGMT methylated vs. unmethylated for LGG data ========\n",
    "        classification_tasks_dict[\"TCGA-LGG_4.301_isMGMTMethylated_base\"]={\n",
    "            \"train_data\": train_data_LGG, \n",
    "            \"test_data_dict\": test_data_LGG_dict, \n",
    "            \"feature_columns\":feature_namelist, \n",
    "            \"keep_feature_columns\": [],\n",
    "            \"label_column\":\"is_MGMT_Methylated\",\n",
    "            \"base_results_path\":base_results_path}\n",
    "\n",
    "        classification_tasks_dict[\"TCGA-LGG_4.302_isMGMTMethylated_withClinicalInfo\"]={\n",
    "            \"train_data\": train_data_LGG, \n",
    "            \"test_data_dict\": test_data_LGG_dict, \n",
    "            \"feature_columns\":feature_namelist,\n",
    "            \"keep_feature_columns\": clinical_namelist,\n",
    "            \"label_column\":\"is_MGMT_Methylated\",\n",
    "            \"base_results_path\":base_results_path}\n",
    "          \n",
    "    else:\n",
    "        raise Exception(\"Unfined experiments!\")\n",
    "      \n",
    "    \n",
    "    for task, classification_settings in classification_tasks_dict.items():\n",
    "        # save the data excel path\n",
    "        classification_settings[\"train_excel_path\"]=train_data_excel_path\n",
    "        classification_settings[\"test_excel_path_dict\"]=test_excel_path_dict\n",
    "        classification_tasks_dict[task]=classification_settings\n",
    "      \n",
    "    return classification_tasks_dict\n",
    "   \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Data preprocessing: Important!!! "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocessing_data(datadf, feature_columns):\n",
    "     #convert the complex data to real data\n",
    "    datadf=convert_complex_to_real(datadf, feature_columns)\n",
    "    \n",
    "    #fill nan values with 0.\n",
    "    #datadf.fillna(value=0, inplace=True)\n",
    "    \n",
    "    return datadf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Main "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    basic_settings=get_basic_settings()\n",
    "    print(\"\\n === basic_settings={} =======\".format(basic_settings))\n",
    "    \n",
    "    # choose the experiment settings.\n",
    "    experiment_class=basic_settings[\"experiment_class\"]\n",
    "    task_list=basic_settings[\"task_list\"]\n",
    "    \n",
    "    # if the task list in the settings is empty, then all the tasks in this experiment class will be done!\n",
    "    if task_list==[]:\n",
    "        classification_tasks_dict=get_classification_tasks_dict(experiment_class, task_name=None)\n",
    "        task_list=list(classification_tasks_dict.keys())\n",
    "    \n",
    "    # Perform the classification task one by one!\n",
    "    for task in task_list:\n",
    "        classification_tasks_dict=get_classification_tasks_dict(experiment_class, task) \n",
    "        classification_task_settings=classification_tasks_dict[task]\n",
    "\n",
    "        #Perform classification\n",
    "        perform_binary_classification(task, classification_task_settings, basic_settings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
